services/node_manager/src/main/java/com/futurewei/alcor/nodemanager/service/implement/NodeServiceImpl.java

```diff
@@ -0,0 +1,146 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+import org.springframework.web.multipart.MultipartFile;
+
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.io.Reader;
+import java.util.ArrayList;
+import java.util.List;
+
+@Service
+public class NodeServiceImpl implements NodeService {
+    private static final Logger logger = LoggerFactory.getLogger(NodeServiceImpl.class);
+
+    @Autowired
+    private NodeRepository nodeRepository;
+
+    public int getNodeInfoFromUpload(MultipartFile file) throws Exception {
+        int nReturn = 0;
+        List<NodeInfo> nodeList = new ArrayList<NodeInfo>();
+        try {
+            Reader reader = new BufferedReader(new InputStreamReader(file.getInputStream()));
+            NodeFileLoader dataCenterConfigLoader = new NodeFileLoader();
+            nodeList = dataCenterConfigLoader.getHostNodeListFromUpload(reader);
+            if (nodeList != null) {
+                nodeRepository.addItemBulkTransaction(nodeList);
+                nReturn = nodeList.size();
+            }
+        } catch (IOException e) {
+            e.printStackTrace();
+        }
+        return nReturn;
+    }
+
+    @Override
+    public int getNodeInfoFromFile(String path) throws Exception {
+        int nReturn = 0;
+        NodeFileLoader dataCenterConfigLoader = new NodeFileLoader();
+        List<NodeInfo> nodeList = dataCenterConfigLoader.loadAndGetHostNodeList(path);
+        if (nodeList != null) {
+            nodeRepository.addItemBulkTransaction(nodeList);
+            nReturn = nodeList.size();
+        }
+        return nReturn;
+    }
+
+    @Override
+    public NodeInfo getNodeInfoById(String nodeId) throws Exception {
+        if (nodeId == null)
+            throw (new ParameterNullOrEmptyException(NodeManagerConstant.NODE_EXCEPTION_PARAMETER_NULL_EMPTY));
+        NodeInfo nodeInfo = null;
+        try {
+            nodeInfo = nodeRepository.findItem(nodeId);
+        } catch (Exception e) {
+            throw e;
+        }
+        return nodeInfo;
+    }
+
+    @Override
+    public List<NodeInfo> getAllNodes() throws Exception {
+        List<NodeInfo> nodes = new ArrayList<NodeInfo>();
+        try {
+            nodes = new ArrayList(nodeRepository.findAllItems().values());
+        } catch (Exception e) {
+            throw e;
+        }
+        return nodes;
+    }
+
+    @Override
+    public NodeInfo createNodeInfo(NodeInfo nodeInfo) throws Exception {
+        if (nodeInfo == null)
+            throw (new ParameterNullOrEmptyException(NodeManagerConstant.NODE_EXCEPTION_PARAMETER_NULL_EMPTY));
+        NodeInfo node = getNodeInfoById(nodeInfo.getId());
+        if (nodeInfo != null) {
+            try {
+                nodeRepository.addItem(nodeInfo);
```

Comment:
We should do all necessary verification before persisting the states to database.

Reply:
Exception for validation and functions to validate data are added.